pinot-segment-local/src/main/java/org/apache/pinot/segment/local/utils/ConsistentDataPushUtils.java

```diff
@@ -0,0 +1,296 @@
+          return true;
+        } catch (SocketTimeoutException se) {
+          // In case of the timeout, we should re-try.
+          return false;
+        } catch (HttpErrorStatusException e) {
+          if (e.getStatusCode() >= 500) {
+            return false;
+          } else {
+            if (e.getStatusCode() == Response.Status.NOT_FOUND.getStatusCode()) {
+              LOGGER.error("Table: {} not found when sending request: {}", rawTableName, startSegmentUri);
+            }
+            throw e;
+          }
+        }
+      });
+    }
+    return uriToLineageEntryIdMap;
+  }
+
+  public static void endReplaceSegments(SegmentGenerationJobSpec spec, Map<URI, String> uriToLineageEntryIdMap) {
+    AuthProvider authProvider = AuthProviderUtils.makeAuthProvider(spec.getAuthToken());
+    String rawTableName = spec.getTableSpec().getTableName();
+    for (URI uri : uriToLineageEntryIdMap.keySet()) {
+      String segmentLineageEntryId = uriToLineageEntryIdMap.get(uri);
+      try {
+        FILE_UPLOAD_DOWNLOAD_CLIENT.endReplaceSegments(
+            FileUploadDownloadClient.getEndReplaceSegmentsURI(uri, rawTableName, TableType.OFFLINE.toString(),
+                segmentLineageEntryId), HttpClient.DEFAULT_SOCKET_TIMEOUT_MS, authProvider);
+      } catch (URISyntaxException e) {
+        throw new RuntimeException("Got invalid controller uri - '" + uri + "'");
+      } catch (HttpErrorStatusException | IOException e) {
+        e.printStackTrace();
+      }
+    }
+  }
+
+  public static void handleUploadException(SegmentGenerationJobSpec spec, Map<URI, String> uriToLineageEntryIdMap,
+      Exception exception) {
+    LOGGER.error("Exception when pushing segments. Marking segment lineage entry to 'REVERTED'.", exception);
+    String rawTableName = spec.getTableSpec().getTableName();
+    for (Map.Entry<URI, String> entry : uriToLineageEntryIdMap.entrySet()) {
+      String segmentLineageEntryId = entry.getValue();
+      try {
+        URI uri =
+            FileUploadDownloadClient.getRevertReplaceSegmentsURI(entry.getKey(), rawTableName, TableType.OFFLINE.name(),
+                segmentLineageEntryId, true);
+        SimpleHttpResponse response = FILE_UPLOAD_DOWNLOAD_CLIENT.revertReplaceSegments(uri);
+        LOGGER.info("Got response {}: {} while sending revert replace segment request for table: {}, uploadURI: {}",
+            response.getStatusCode(), response.getResponse(), rawTableName, entry.getKey());
+      } catch (URISyntaxException | HttpErrorStatusException | IOException e) {
+        LOGGER.error("Exception when sending revert replace segment request to controller: {} for table: {}",
+            entry.getKey(), rawTableName, e);
+      }
+    }
+  }
+
+  /**
+   * Ensures that all files in tarFilePaths have the expected tar file extension and strip the extension to obtain
+   * segment names.
+   */
+  public static List<String> getTarSegmentsTo(List<String> tarFilePaths) {
```

Comment:
This function is very similar to getMetadataSegmentsTo() can we fold them and pass the key set directly to the other one?